### PR TITLE
Extract shared CRUD pattern from admin notification channels (#80)

### PR DESCRIPTION
**Pull request by dpage**
## Summary

- Extract shared CRUD state management from `AdminEmailChannels` (983→417 lines) and `AdminWebhookChannels` (1015→350 lines) into a reusable `useChannelCRUD` hook, `ChannelTable`, and `ChannelDialogShell` components.
- Decompose each channel component into focused sub-components (settings, recipients, headers, auth, templates tabs).
- Fix `key={index}` anti-pattern in webhook headers by generating stable UUIDs for header entries.
- Add 168 new tests across 7 test files covering shared infrastructure, both channel orchestrators, webhook helpers, and all edge cases.

### Before/After Line Counts

| File | Before | After |
|------|--------|-------|
| AdminEmailChannels.tsx | 983 | 417 |
| AdminWebhookChannels.tsx | 1,015 | 350 |

### New Shared Components

| File | Lines | Purpose |
|------|-------|---------|
| channels/useChannelCRUD.ts | 266 | Shared CRUD hook (fetch, toggle, test, delete, dialog) |
| channels/ChannelTable.tsx | 262 | Shared channel list table |
| channels/ChannelDialogShell.tsx | 112 | Shared dialog wrapper with tabs |
| channels/channelTypes.ts | 26 | BaseChannel, ChannelColumnDef types |

### Coverage

All new/modified files meet the 90% line coverage floor:
- useChannelCRUD.ts: 100%
- ChannelTable.tsx: 100%
- ChannelDialogShell.tsx: 100%
- AdminEmailChannels.tsx: 91.36%
- AdminWebhookChannels.tsx: 90.28%
- webhookHelpers.ts: 100%

## Test plan

- [x] All 168 new tests pass
- [x] Existing AdminPanel tests pass (AdminTokenScopes, AdminProbes — 24 tests)
- [x] Hook tests pass (237 tests)
- [x] Utils tests pass (318 tests)
- [x] ESLint: zero errors on all new/modified files
- [x] TypeScript: no new type errors introduced
- [ ] CI pipeline passes
- [ ] CodeRabbit review addressed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * New reusable channel table and tabbed dialog shell; added email recipients/settings and webhook settings/headers/auth/templates tabs and recipient-count display.
  * New shared channel types, form defaults, and webhook helper utilities for auth/headers/templates.

* **Refactor**
  * Centralized channel CRUD/dialog state and flows for email and webhook panels with unified loading, saving, error, success, and pending-recipient handling.

* **Tests**
  * Expanded unit and UI tests covering CRUD flows, toggles, test-send, recipients/headers/auth/templates, and the shared CRUD hook.

### CODE REVIEW

**Comment by dpage** [comment]
Addressed all CodeRabbit findings in 0957419:

**Actionable (Major):**
1. **Duplicate channel creation risk** — Fixed: dialog now closes immediately after channel creation succeeds. Pending recipients are provisioned as a best-effort follow-up with their own try/catch. If any recipients fail, a warning message is shown without reverting the channel.
2. **Stale recipient responses** — Fixed: added `editingChannelIdRef` to track the active channel. All async recipient operations (fetch, add, toggle, delete) now check the ref after their await before committing state updates. Late/stale responses are silently discarded.

**Nitpicks:**
3. **Mask secret credentials** — Fixed: Token and API Key Value fields now use `type="password"` and `autoComplete="off"`.
4. **JSDoc for mapChannel** — Fixed: added warning that `mapChannel` must be stable (module scope or memoized) to prevent infinite re-fetch loops.
5. **Email validation** — Fixed: `handleAddRecipient` now checks for `@` in the email. Add button is disabled when email lacks `@`.

All fixes include corresponding test coverage (182 tests passing).

File: client/src/components/AdminPanel/__tests__/AdminEmailChannels.test.tsx [hunk 1]
```diff
@@ -0,0 +1,1243 @@
...
+                        name: 'Updated Email',
+                    })
+                );
+            });
+
+            // Should NOT include unchanged fields
+            const putCall = mockApiPut.mock.calls[0];
+            expect(putCall[1]).not.toHaveProperty('smtp_host');
+            expect(putCall[1]).not.toHaveProperty('smtp_port');
+        });
```
[the remark this answers is not in the document]
Comment (dpage) [reply client/src/components/AdminPanel/__tests__/AdminEmailChannels.test.tsx:372]: The component does send only changed fields in edit mode — see lines 175-211 of AdminEmailChannels.tsx which build a diff body by comparing each form field against `crud.editingChannel`. The test assertion is correct.

File: client/src/components/AdminPanel/__tests__/AdminEmailChannels.test.tsx [hunk 2]
```diff
@@ -0,0 +1,1243 @@
...
+
+            await waitFor(() => {
+                expect(mockApiPost).toHaveBeenCalledWith(
+                    '/api/v1/notification-channels/1/recipients',
+                    expect.objectContaining({
+                        email_address: 'newuser@example.com',
+                        enabled: true,
+                    })
+                );
+            });
```
[the remark this answers is not in the document]
Comment (dpage) [reply client/src/components/AdminPanel/__tests__/AdminEmailChannels.test.tsx:745]: The component does send `enabled: true` — see line 314 of AdminEmailChannels.tsx where `handleAddRecipient` posts `{ email_address: email, display_name: name, enabled: true }`. The test assertion matches the actual behavior.

File: client/src/components/AdminPanel/AdminEmailChannels.tsx [hunk 3]
```diff
@@ -8,187 +8,78 @@
...
+    // Ref to track the currently-editing channel ID for staleness checks.
+    // This avoids race conditions when the user quickly switches channels.
+    const editingChannelIdRef = useRef<number | null>(null);
 
+    // Keep ref in sync with crud.editingChannel
     useEffect(() => {
-        fetchChannels();
-    }, [fetchChannels]);
+        editingChannelIdRef.current = crud.editingChannel?.id ?? null;
+    }, [crud.editingChannel]);
```
[the remark this answers is not in the document]
Comment (dpage) [reply client/src/components/AdminPanel/AdminEmailChannels.tsx:71]: Fixed in df489a2: `editingChannelIdRef.current` is now set synchronously in `handleOpenCreate` (null), `handleOpenEdit` (channel.id), and `handleCloseDialog` (null) before any async operations. The useEffect is retained as a safety net.

File: client/src/components/AdminPanel/channels/useChannelCRUD.ts [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        } catch (err: unknown) {
+            if (err instanceof Error) {
+                setError(err.message);
+            } else {
+                setError('An unexpected error occurred');
+            }
+        } finally {
+            setLoading(false);
+        }
+    }, [channelType, mapChannel]);
```
[the remark this answers is not in the document]
Comment (dpage) [reply client/src/components/AdminPanel/channels/useChannelCRUD.ts:149]: Fixed in df489a2: added `fetchRequestIdRef` counter. Each call to `fetchChannels` captures a request ID, and `setChannels`/`setError`/`setLoading` are guarded by checking that the request ID still matches the latest.

File: client/src/components/AdminPanel/AdminEmailChannels.tsx [hunk 5]
```diff
@@ -8,187 +8,78 @@
...
     const fetchRecipients = useCallback(async (channelId: number) => {
         try {
             setRecipientsLoading(true);
             const data = await apiGet<Record<string, unknown>>(
                 `/api/v1/notification-channels/${channelId}/recipients`
             );
+            // Guard against stale responses: only update if still editing same channel
+            if (editingChannelIdRef.current !== channelId) { return; }
             const raw = (data.recipients || data || []) as Record<string, unknown>[];
             const mapped: EmailRecipient[] = raw.map(
```
[the remark this answers is not in the document]
Comment (dpage) [reply client/src/components/AdminPanel/AdminEmailChannels.tsx:88]: Fixed in 5ce46d2: recipients data is now guarded with `Array.isArray` checks before calling `.map()`.

File: client/src/components/AdminPanel/AdminEmailChannels.tsx [hunk 6]
```diff
@@ -240,88 +139,82 @@ const AdminEmailChannels: React.FC = () => {
...
-            setDialogError('SMTP port must be a valid port number (1-65535).');
+            crud.setDialogError('SMTP port must be a valid port number (1-65535).');
             return;
         }
 
         try {
-            setSaving(true);
-            setDialogError(null);
+            crud.setSaving(true);
+            crud.setDialogError(null);
```
[the remark this answers is not in the document]
Comment (dpage) [reply client/src/components/AdminPanel/AdminEmailChannels.tsx:182]: Fixed in 5ce46d2: `handleSaveChannel` now returns early if `recipientSaving` is true, save button is disabled during recipient mutations, and `setRecipientSaving(false)` is called unconditionally in all `finally` blocks to prevent stuck state.
